KEYCLOAK-11770 add an hardcoded attribute mapper for ldap provider

micedre opened the pull request:
Added a new hardcoded attribute mapper for ldap provider. This allows to hardcode any attribute and/or the emailVerified or enabled property on ldap user, even when the ldap provider is not set to import user.

<!---
Please read https://github.com/keycloak/keycloak/blob/master/CONTRIBUTING.md and follow these guidelines when contributing to Keycloak
-->

micedre committed:
KEYCLOAK-11770 add an hardcoded attribute mapper

Signed-off-by: Cédric Couralet <cedric.couralet@insee.fr>

mposolda commented on federation/ldap/src/main/java/org/keycloak/storage/ldap/mappers/HardcodedAttributeMapper.java:
```diff
@@ -0,0 +1,228 @@
+    public static final String USER_MODEL_ATTRIBUTE = "user.model.attribute";
+    public static final String LDAP_ATTRIBUTE = "ldap.attribute";
+    public static final String ATTRIBUTE_VALUE = "attribute.value";
+    public static final String READ_ONLY = "read.only";
```
It seems many of those constants are not used in the mapper. Could you please remove unused constants?

mposolda commented on federation/ldap/src/main/java/org/keycloak/storage/ldap/mappers/HardcodedAttributeMapper.java:
```diff
@@ -0,0 +1,228 @@
+    private static final Map<String, Property<Object>> userModelProperties;
+
+    static {
+        Map<String, Property<Object>> userModelProps = PropertyQueries.createQuery(UserModel.class)
```
NOTE: It seems that same userModelProps stuff is used in some other mappers (EG. UserAttributeLdapMapper). In that case I suggest to make the field "userModelProperties" on UserAttributeLDAPMapper public and reference it from this class to avoid the same static code block duplicated in both classes. Even better could be to move this block to some util class and make both mappers to reference to it, but up to you :)

mposolda commented on federation/ldap/src/main/java/org/keycloak/storage/ldap/mappers/HardcodedAttributeMapper.java:
```diff
@@ -0,0 +1,228 @@
+      return mapperModel.getConfig().getFirst(ATTRIBUTE_VALUE);
+   }
+
+   private boolean isBinaryAttribute() {
```
Same apply for those private methods, which are not used anymore. I understand you copied the implementation from some other class, but please cleanup unused stuff.

mposolda commented on federation/ldap/src/main/java/org/keycloak/storage/ldap/mappers/HardcodedAttributeMapper.java:
```diff
@@ -0,0 +1,228 @@
+        Property<Object> userModelProperty = userModelProperties.get(userModelAttrName.toLowerCase());
+
+        if (userModelProperty != null) {
+            checkDuplicateEmail(userModelAttrName, attributeValue, realm, ldapProvider.getSession(), user);
```
Does it have any sense to check for duplicated email here? IMO I am not seeing much sense to have hardcoded username or email same for all users? So I would just delete this line and also completely delete the methods "checkDuplicateEmail" and "checkDuplicateUsername" . And you can instead implement method on the HardcodedAttributeMapperFactory "validateConfiguration" and throw an error in case that "username" or "email" was configured as user model attribute.

mposolda reviewed (changes requested):
@micedre Thanks for the PR! Added few comments to it.

micedre commented:
@mposolda thx for the review, I cleaned up unused code, put the static block in the `LDAPUtils` class (not sure if it the right place), and added some checks for the configuration.

mposolda commented:
@micedre Thanks for the updates. This PR is good from the code perspective.

So one additional thing would be to add the documentation for this. You may need to send the PR to "keycloak/keycloak-documentation" for this and add the paragraph with the new mapper similar to the paragraph with the existing mappers. Thanks!

mposolda reviewed (approved):
@micedre Thanks!

tkyjovsk commented:
@mposolda It seems that merging this PR introduced a compilation error. The class `LDAPHardcodedAttributeTest` is using class `RunOnServerDeployment` which has been recently removed as part of KEYCLOAK-12167, PR: https://github.com/keycloak/keycloak/pull/6514
